packages/apollo-server/src/utils/types.ts

```diff
@@ -0,0 +1,90 @@
+import { GraphQLSchema } from 'graphql';
+import { CorsOptions } from 'cors';
+import { SchemaDirectiveVisitor, IResolvers } from 'graphql-tools';
+import { ConnectionContext } from 'subscriptions-transport-ws';
+import { GraphQLOptions } from 'apollo-server-core';
+
+export type Context<T = any> = T;
+export type ContextFunction<T = any> = (
+  context: Context<T>,
+) => Promise<Context<T>>;
+
+export interface SubscriptionServerOptions {
+  path?: string;
+  onConnect?: Function;
+  onDisconnect?: Function;
+  keepAlive?: number;
+}
+
+export interface Config<Server, ContextShape = any, Cors = CorsOptions>
+  extends Pick<
+      GraphQLOptions<Context<ContextShape>>,
+      | 'formatError'
+      | 'debug'
+      | 'rootValue'
+      | 'logFunction'
+      | 'formatParams'
+      | 'validationRules'
+      | 'formatResponse'
+      | 'fieldResolver'
+      | 'debug'
+      | 'cacheControl'
+      | 'tracing'
+    > {
+  app?: Server;
+  typeDefs?: string | [string];
+  resolvers?: IResolvers;
+  schema?: GraphQLSchema;
+  schemaDirectives: Record<string, typeof SchemaDirectiveVisitor>;
+  context?: Context<ContextShape> | ContextFunction<ContextShape>;
+  cors?: Cors;
```

Comment:
So we are comfortable assuming that the api of the npm cors module can be implemented for every potential variant?

Reply:
Actually it's already the case that the `cors` option for koa has different semantics than for express. The options just have different names. (Maybe some of the most basic usage is in common.)

Reply:
implementing classes can override this generic type? (see the koa one)

Reply:
Yeah, I guess it makes sense in TypeScript terms, I guess the docs are just gonna be a little weird, like "cors is the argument for configuring a cors module, either `cors` or `@koa/cors` or ...)